numba/targets/arrayobj.py

```diff
@@ -3756,70 +3756,83 @@ def take_impl(a, indices):
-def numpy_arange_2(context, builder, sig, args):
-    dtype = as_dtype(sig.return_type.dtype)
+@overload(np.arange)
+def np_arange(start, stop=None, step=None, dtype=None):
+    if isinstance(stop, types.Optional):
+        stop = stop.type
+    if isinstance(step, types.Optional):
+        step = step.type
+    if isinstance(dtype, types.Optional):
+        dtype = dtype.type
 
-    def arange(start, stop):
-        return np.arange(start, stop, 1, dtype)
+    if stop is None:
+        stop = types.none
+    if step is None:
+        step = types.none
+    if dtype is None:
+        dtype = types.none
 
-    res = context.compile_internal(builder, arange, sig, args)
-    return impl_ret_new_ref(context, builder, sig.return_type, res)
+    if (not isinstance(start, types.Number) or
+        not isinstance(stop, (types.NoneType, types.Number)) or
+        not isinstance(step, (types.NoneType, types.Number)) or
+            not isinstance(dtype, (types.NoneType, types.DTypeSpec))):
 
+        return
 
-@lower_builtin(np.arange, types.Number, types.Number,
-               types.Number)
-def numpy_arange_3(context, builder, sig, args):
-    dtype = as_dtype(sig.return_type.dtype)
+    if isinstance(dtype, types.NoneType):
+        true_dtype = _arange_dtype(start, stop, step)
+    else:
+        true_dtype = dtype.dtype
 
-    def arange(start, stop, step):
-        return np.arange(start, stop, step, dtype)
+    use_complex = any([isinstance(x, types.Complex)
+                       for x in (start, stop, step)])
 
-    res = context.compile_internal(builder, arange, sig, args)
-    return impl_ret_new_ref(context, builder, sig.return_type, res)
+    start_value = getattr(start, "literal_value", None)
+    stop_value = getattr(stop, "literal_value", None)
+    step_value = getattr(step, "literal_value", None)
 
+    def impl(start, stop=None, step=None, dtype=None):
+        # Allow for improved performance if given literal arguments.
+        lit_start = start_value if start_value is not None else start
+        lit_stop = stop_value if stop_value is not None else stop
+        lit_step = step_value if step_value is not None else step
 
-@lower_builtin(np.arange, types.Number, types.Number,
-               types.Number, types.DTypeSpec)
-def numpy_arange_4(context, builder, sig, args):
+        _step = lit_step if lit_step is not None else 1
+        if lit_stop is None:
+            _start, _stop = 0, lit_start
+        else:
+            _start, _stop = lit_start, lit_stop
 
-    if any(isinstance(a, types.Complex) for a in sig.args):
-        def arange(start, stop, step, dtype):
-            nitems_c = (stop - start) / step
-            nitems_r = math.ceil(nitems_c.real)
-            nitems_i = math.ceil(nitems_c.imag)
+        nitems_c = (_stop - _start) / _step
+        nitems_r = int(math.ceil(nitems_c.real))
+
+        # Binary operator needed for compiler branch pruning.
+        if use_complex == True: # noqa 712
```

Comment:
Apologies, I'm not sure if I made myself clear. Did the `if use_complex is True` not work with branch pruning? I'd prefer that over a `flake8` ignore and use of `==`.

Reply:
How can we check this?  I'm not sure how to do an `inspect_types` on an overloaded function.

Reply:
You can do `NUMBA_DEBUG=1`, however, I don't think the code in its present state would make it through type inference if such a change broke branch pruning.